### PR TITLE
ocamlyacc: support OCaml-style comments

### PR DESCRIPTION
**Pull request by DemiMarie**
This adds support for OCaml-style comments everywhere in ocamlyacc, in addition to the existing support for C-style comments.  The syntax is the same as for OCaml comments in actions.  It also fixes reporting of an unterminated comment in an action, which was previously reported as `unmatched /*` instead of `unmatched (*`

### CODE REVIEW

**Comment by shindere** [comment]
I like the idea, thanks. I did a shallow review only.

I notice you added the test for comments to an existing test. Wouldn't
it be better to leave that test as it is and introduce another,
dedicated test?

**Comment by DemiMarie** [comment]
> I like the idea, thanks.

You’re welcome!

> I did a shallow review only. I notice you added the test for comments to an existing test. Wouldn't it be better to leave that test as it is and introduce another, dedicated test?

I don’t think it makes a difference here.

File: yacc/reader.c [hunk 1]
```diff
@@ -199,7 +199,8 @@ static void process_quoted_string(char c, FILE *const f)
     for (;;)
     {
         c = *cptr++;
-        putc(c, f);
+        if (f)
+            putc(c, f);
```
Comment (damiendoligez) [inline yacc/reader.c]: - Is this relevant to the change in comment syntax?
- Do you have a case where this function is called with f==NULL?
- Is it appropriate to do nothing in this case as opposed to signaling an error?

**Review by damiendoligez — COMMENTED** [review]
Looks good except for all these new conditions `if (f) putc...` and `if (f) fwrite...`. If this is really needed, it would be better to define and use  two functions `safe_putc` and `safe_fwrite`.

File: yacc/reader.c (hunk 1 above)
Comment (DemiMarie) [reply yacc/reader.c]: It is, as OCaml comments can contain quoted strings.

**Review by damiendoligez — APPROVED** [review]
Looks good now. Can you resolve the conflict and add my name as reviewer ? I'll merge after that.

**Comment by DemiMarie** [comment]
@damiendoligez done.

**Comment by ferminr** [comment]
I think the manual could be updated to reflect the new syntax for comments. Specifically, where it says:

Comments are enclosed between `/*` and `*/` (as in C) in the “declarations” and “rules” sections, and between `(*` and `*)` (as in OCaml) in the “header” and “trailer” sections.

**Comment by damiendoligez** [comment]
@DemiMarie Could you update the manual?

**Comment by damiendoligez** [comment]
Thanks !

**Comment by sidkshatriya** [comment]
I think these changes may have led to a small regression.

When I do `dune build @libs` in ocaml source code now, I get:

```
$ dune build @libs
File "ocamltest/tsl_parser.mly", line 1: syntax error
(**************************************************************************)
^
```

I don't know how important this error is and if it has any practical impact.

**Comment by DemiMarie** [comment]
@sidkshatriya That means you are using a version of ocamlyacc that isn’t the one in the OCaml source tree.  Not sure what the best fix is.

**Comment by sidkshatriya** [comment]
Thanks for your very quick reply ! I'm sorry I'm not very familiar with ocamlyacc.

If you revert to a recent commit (e.g. use fd6e6e0c9cf636f69cc90f4590834f77efaa5b95 ), invoking `dune build @libs` does not result in error mentioned above.

I am compiling ocaml in a fresh OCaml 5.0.0 switch. I only have done `opam install dune`. That's it.

**Comment by gasche** [comment]
The dune build system for the OCaml compiler is morally a hack, it uses the tools from the ambient opam switch to build the compiler codebase -- which can break whenever the compiler distribution has more recent tools than the opam switch, and uses the more recent features in its own codebase.

This is what is going on here: @DemiMarie introduced a new ocamlyacc feature, and also started using it in the compiler codebase in the non-essential commit 514a830eae3991aec12146d3f6aeff3aa2d8406a.
If we want dune to work well, a reasonable solution is to revert 514a830eae3991aec12146d3f6aeff3aa2d8406a for now, and wait until the next OCaml version is released (with the new ocamlyacc feature) to merge it again. At this point the dune build will be possible with the most recent OCaml release as your opam switch.

(I think that having dune more-or-less working to build subset of the compiler distribution is more useful than having a consistent comment syntax for .mly files, so I would suggest doing that.)
